src/portable/portable.c

```diff
@@ -74,17 +75,24 @@ static bool unit_match(const char *unit, char **matches) {
         return false;
 }
 
-static PortableMetadata *portable_metadata_new(const char *name, int fd) {
-        PortableMetadata *m;
+static PortableMetadata *portable_metadata_new(const char *name, const char *path, int fd) {
+        _cleanup_(portable_metadata_unrefp) PortableMetadata *m;
 
         m = malloc0(offsetof(PortableMetadata, name) + strlen(name) + 1);
         if (!m)
                 return NULL;
 
+        /* In case of a layered attach, we want to remember which image the unit came from */
+        if (path) {
+                m->image_path = strdup(path);
+                if (!m->image_path)
+                        return NULL;
```

Comment:
the error path is hosed here. you initialized the object to all zeroes via malloc0(), i.e. `m->fd` is going to be 0, which is a valid fd, and thus if this error path i hit we'l end up closing STDIN by accident. So you either have to initialize m->fd to -1 early on, or you drop the use of `_cleanup_` here for the object, and just insert a regular free() into this error path.

Reply:
ouch! good catch, removed _cleanup_